src/core/net/srp_client.cpp

```diff
@@ -1500,51 +1503,86 @@ void Client::EnableAutoStartMode(AutoStartCallback aCallback, void *aContext)
-        // the list of servers. As we iterate through server entries,
-        // with probability `1/numServers`, we choose to switch the
-        // current selected server with the new entry. This approach
-        // results in a uniform/same probability of selection among
-        // all server entries.
-
-        if ((numServers == 1) || (Random::NonCrypto::GetUint16InRange(0, numServers) == 0))
+        if (IsRunning() && mAutoStartDidSelectServer && mAutoStartIsUsingAnycastAddress)
         {
-            selectedServer = server;
+            if ((mServerInstanceNumber == anycastInfo.mSerialNumber) &&
+                (GetServerAddress().GetAddress() == anycastInfo.mAnycastAddress))
+            {
+                // Client is already using the same anycast address.
+                ExitNow();
+            }
         }
 
-        if (IsRunning() && mAutoStartDidSelectServer && (GetServerAddress() == server.mSockAddr))
+        otLogInfoSrp("[client] Found anycast server %d", anycastInfo.mSerialNumber);
+
+        serverSockAddr.SetAddress(anycastInfo.mAnycastAddress);
+        serverSockAddr.SetPort(mAnycastServerPort);
+        mServerInstanceNumber = anycastInfo.mSerialNumber;
+        serverIsAnycast       = true;
+    }
+    else
+    {
+        uint16_t                                  numServers = 0;
+        NetworkData::Service::DnsSrpUnicast::Info unicastInfo;
+        NetworkData::Service::Manager::Iterator   iterator;
+
+        while (Get<NetworkData::Service::Manager>().GetNextDnsSrpUnicastInfo(iterator, unicastInfo) == kErrorNone)
         {
-            ExitNow();
+            if (IsRunning() && mAutoStartDidSelectServer && !mAutoStartIsUsingAnycastAddress &&
+                (GetServerAddress() == unicastInfo.mSockAddr))
+            {
+                ExitNow();
+            }
+
+            numServers++;
+
+            // Choose a server randomly (with uniform distribution) from
+            // the list of servers. As we iterate through server entries,
+            // with probability `1/numServers`, we choose to switch the
+            // current selected server with the new entry. This approach
+            // results in a uniform/same probability of selection among
+            // all server entries.
+
+            if ((numServers == 1) || (Random::NonCrypto::GetUint16InRange(0, numServers) == 0))
+            {
+                serverSockAddr  = unicastInfo.mSockAddr;
+                serverIsAnycast = false;
+            }
         }
     }
 
     if (IsRunning())
     {
-        otLogInfoSrp("[client] Server %s is no longer present in net data", GetServerAddress().ToString().AsCString());
         Stop(kRequesterAuto);
     }
 
-    VerifyOrExit(numServers > 0);
-    IgnoreError(Start(selectedServer.mSockAddr, kRequesterAuto));
+    VerifyOrExit(!serverSockAddr.GetAddress().IsUnspecified());
+
+    mAutoStartIsUsingAnycastAddress = serverIsAnycast;
```

Comment:
What is the benefit of tracking `mAutoStartIsUsingAnycastAddress`?
Can we just update the current srp server and port to whatever determined from `Network Data`?

Reply:
Thanks for clarification.